### PR TITLE
coprocessor, dag: add arithmetic operations.

### PR DESCRIPTION
**Pull request by hicqu**

### CODE REVIEW

File: src/coprocessor/dag/expr/arithmetic.rs [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+}
+
+#[inline]
+fn do_arithmetic<T, F>(lhs: Option<T>, rhs: Option<T>, op: F) -> Result<Option<T>>
+where
+    F: Fn(T, T) -> Result<T>,
+{
+    match (lhs, rhs) {
+        (None, _) | (_, None) => Ok(None),
+        (Some(lhs), Some(rhs)) => op(lhs, rhs).map(|t| Some(t)),
```
Comment (andelf) [inline src/coprocessor/dag/expr/arithmetic.rs]: make this first match arm, saves comparison times.

**Comment by AndreMouche** [comment]
Please fix the CI

File: src/coprocessor/dag/expr/arithmetic.rs [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+
+    pub fn minus_decimal<'a, 'b: 'a>(
+        &'b self,
+        ctx: &StatementContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Decimal>>> {
+        let lhs = try!(self.children[0].eval_decimal(ctx, row));
+        let rhs = try!(self.children[1].eval_decimal(ctx, row));
+        do_arithmetic(lhs, rhs, |l, r| {
+            let result: Result<Decimal> = (l.as_ref() - r.as_ref()).into();
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/arithmetic.rs]: Should `ignore_truncate` be checked?

File: src/coprocessor/dag/expr/arithmetic.rs [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+            result.map(Cow::Owned)
+        })
+    }
+
+    pub fn minus_int(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        let lhs = try!(self.children[0].eval_int(ctx, row));
+        let rhs = try!(self.children[1].eval_int(ctx, row));
+        do_arithmetic(lhs, rhs, |l, r| l.checked_sub(r).ok_or(Error::Overflow))
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/arithmetic.rs]: How to deal with `i64::MAX as u64 - (-1i64)`?

File: src/coprocessor/dag/expr/arithmetic.rs [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+            res.map(|u| u as i64)
+        })
+    }
+
+    pub fn multiply_real(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<f64>> {
+        let lhs = try!(self.children[0].eval_real(ctx, row));
+        let rhs = try!(self.children[1].eval_real(ctx, row));
+        do_arithmetic(lhs, rhs, |l, r| {
+            let res = l * r;
+            if !res.is_finite() {
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/arithmetic.rs]: Should it return `INF` or an error?

File: src/coprocessor/dag/expr/arithmetic.rs [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+
+    pub fn multiply_decimal<'a, 'b: 'a>(
+        &'b self,
+        ctx: &StatementContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Decimal>>> {
+        let lhs = try!(self.children[0].eval_decimal(ctx, row));
+        let rhs = try!(self.children[1].eval_decimal(ctx, row));
+        do_arithmetic(lhs, rhs, |l, r| {
+            let result: Result<Decimal> = (l.as_ref() * r.as_ref()).into();
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/arithmetic.rs]: I believe the error is dropped in tidb's implementation.

File: src/coprocessor/dag/expr/arithmetic.rs [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+            result.map(Cow::Owned)
+        })
+    }
+
+    pub fn multiply_int(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        let lhs = try!(self.children[0].eval_int(ctx, row));
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/arithmetic.rs]: How to deal with `1u64 * -1i61`?

File: src/coprocessor/dag/expr/arithmetic.rs [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+
+    pub fn plus_decimal<'a, 'b: 'a>(
+        &'b self,
+        ctx: &StatementContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Decimal>>> {
+        let lhs = try!(self.children[0].eval_decimal(ctx, row));
+        let rhs = try!(self.children[1].eval_decimal(ctx, row));
+        do_arithmetic(lhs, rhs, |l, r| {
+            let result: Result<Decimal> = (l.as_ref() + r.as_ref()).into();
```
Comment (hicqu) [inline src/coprocessor/dag/expr/arithmetic.rs]: I think we should it return `Result`, not `Decimal`. Right?

File: src/coprocessor/dag/expr/arithmetic.rs (hunk 2 above)
Comment (hicqu) [inline src/coprocessor/dag/expr/arithmetic.rs]: In TiDB, it seems like we don't need to consider that.

File: src/coprocessor/dag/expr/arithmetic.rs (hunk 4 above)
Comment (hicqu) [inline src/coprocessor/dag/expr/arithmetic.rs]: In TiDB:
```go
if math.IsInf(result, 0) {
        return 0, true, types.ErrOverflow.GenByArgs("DOUBLE", fmt.Sprintf("(%s * %s)", s.args[0].String(), s.args[1].String()))
}
```
So I thinks it's OK.

File: src/coprocessor/dag/expr/arithmetic.rs (hunk 5 above)
Comment (hicqu) [inline src/coprocessor/dag/expr/arithmetic.rs]: Really? I find this:
```go
    err = types.DecimalMul(a, b, c)
    if err != nil {
        return nil, true, errors.Trace(err)
    }
```

File: src/coprocessor/dag/expr/arithmetic.rs (hunk 6 above)
Comment (hicqu) [inline src/coprocessor/dag/expr/arithmetic.rs]: you mean  `1u64 * -1i64`?  TiDB will deal with this case by converting both of them to `u64` or `i64`. So We won't get `u64 * i64` or `i64 * u64` in any signatures. We only have `MultiplyInt` and `MultiplyIntUnsigned`, which deal with `i64` and `u64` respectively.

File: src/coprocessor/dag/expr/arithmetic.rs (hunk 3 above)
Comment (hicqu) [inline src/coprocessor/dag/expr/arithmetic.rs]: TiDB converts both 2 arguments to unsigned, and tell TiKV use **MinusIntUnsigned** to calculate it. So `i64::MAX as u64 - (-1)`  won't cause any failure, but `u64::MAX - (-1)` will cause out-of-range failure.

File: src/coprocessor/dag/expr/arithmetic.rs (hunk 4 above)
Comment (hicqu) [inline src/coprocessor/dag/expr/arithmetic.rs]: In MySQL, if the result is infinate, raises error.

**Comment by hicqu** [comment]
@BusyJay , PTAL, thanks.

File: src/coprocessor/dag/expr/arithmetic.rs [hunk 8]
```diff
@@ -0,0 +1,401 @@
...
+        let lhs = try_opt!(self.children[0].eval_int(ctx, row));
+        let rhs = try_opt!(self.children[1].eval_int(ctx, row));
+        let lus = mysql::has_unsigned_flag(self.children[0].get_tp().get_flag());
+        let rus = mysql::has_unsigned_flag(self.children[1].get_tp().get_flag());
+        let res = match (lus, rus) {
+            (true, true) => try!((lhs as u64).checked_sub(rhs as u64).ok_or(Error::Overflow)),
+            (true, false) => if rhs >= 0 {
+                try!((lhs as u64).checked_sub(rhs as u64).ok_or(Error::Overflow))
+            } else {
+                let rhs = try!(rhs.checked_neg().ok_or(Error::Overflow)) as u64;
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/arithmetic.rs]: You can use `rhs.checked_neg().and_then(|rhs| (rhs as u64).checked_add(lhs as u64))`. And change L109 to `res.ok_or(Error::Overflow).map(Some)`.

File: src/coprocessor/dag/expr/arithmetic.rs [hunk 9]
```diff
@@ -0,0 +1,401 @@
...
+        let lhs = try_opt!(self.children[0].eval_int(ctx, row));
+        let rhs = try_opt!(self.children[1].eval_int(ctx, row));
+        let lus = mysql::has_unsigned_flag(self.children[0].get_tp().get_flag());
+        let rus = mysql::has_unsigned_flag(self.children[1].get_tp().get_flag());
+
+        let (mut a_neg, mut b_neg) = (false, false);
+        let (mut l, mut r) = (lhs as u64, rhs as u64);
+        if !lus && lhs < 0 {
+            a_neg = true;
+            l = Wrapping(lhs).neg().0 as u64;
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/arithmetic.rs]: Use `opp_neg!` instead.

File: src/coprocessor/dag/expr/arithmetic.rs [hunk 10]
```diff
@@ -0,0 +1,401 @@
...
+        }
+        if !rus && rhs < 0 {
+            b_neg = true;
+            r = Wrapping(rhs).neg().0 as u64;
+        }
+        let res = try!(l.checked_mul(r).ok_or(Error::Overflow));
+        if a_neg != b_neg {
+            try!((i64::MAX as u64).checked_sub(res).ok_or(Error::Overflow));
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/arithmetic.rs]: Compare it directly.

File: src/coprocessor/dag/expr/arithmetic.rs [hunk 11]
```diff
@@ -0,0 +1,401 @@
...
+        let lhs = try_opt!(self.children[0].eval_int(ctx, row));
+        let rhs = try_opt!(self.children[1].eval_int(ctx, row));
+        let lus = mysql::has_unsigned_flag(self.children[0].get_tp().get_flag());
+        let rus = mysql::has_unsigned_flag(self.children[1].get_tp().get_flag());
+        let res = match (lus, rus) {
+            (true, true) => try!((lhs as u64).checked_add(rhs as u64).ok_or(Error::Overflow)),
+            (true, false) => if rhs >= 0 {
+                try!((lhs as u64).checked_add(rhs as u64).ok_or(Error::Overflow))
+            } else {
+                let rhs = try!(rhs.checked_neg().ok_or(Error::Overflow)) as u64;
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/arithmetic.rs]: I don't think checked_neg is correct here. For example, `i64::MAX + 1 + i64::MIN` should return 0. Use `opp_neg!` instead.

File: src/coprocessor/dag/expr/arithmetic.rs [hunk 12]
```diff
@@ -0,0 +1,407 @@
...
+    }
+
+    pub fn plus_decimal<'a, 'b: 'a>(
+        &'b self,
+        ctx: &StatementContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Decimal>>> {
+        let lhs = try_opt!(self.children[0].eval_decimal(ctx, row));
+        let rhs = try_opt!(self.children[1].eval_decimal(ctx, row));
+        let result: Result<Decimal> = (lhs.as_ref() + rhs.as_ref()).into();
```
Comment (hicqu) [inline src/coprocessor/dag/expr/arithmetic.rs]: It's hard to pass compiler so I prefer to keep it.

Comment (BusyJay) [inline src/coprocessor/dag/expr/arithmetic.rs]: Why?

Comment (hicqu) [inline src/coprocessor/dag/expr/arithmetic.rs]: lhs and rhs are both `Cow<Decimal>`. We can't use `lhs.add(&rhs)` directly. `as_ref` is still needed.

Comment (BusyJay) [inline src/coprocessor/dag/expr/arithmetic.rs]: Have you tried it or just a guess?

Comment (hicqu) [inline src/coprocessor/dag/expr/arithmetic.rs]: I have tried.
```
$ cargo build
   Compiling tikv v0.0.1 (file:///home/qupeng/github.com/pingcap/tikv)
error[E0599]: no method named `add` found for type `alloc::borrow::Cow<'_, coprocessor::codec::mysql::decimal::Decimal>` in the current scope
  --> /home/qupeng/github.com/pingcap/tikv/src/coprocessor/dag/expr/arithmetic.rs:39:43
   |
39 |         let result: Result<Decimal> = lhs.add(&rhs).into();
   |                                           ^^^
   |
   = help: items from traits can only be used if the trait is in scope
   = note: the following trait is implemented but not in scope, perhaps add a `use` for it:
           candidate #1: `use std::ops::Add;`
```

Comment (BusyJay) [inline src/coprocessor/dag/expr/arithmetic.rs]: As the error message says, you need to import trait `Add` to use method `add`.

Comment (hicqu) [inline src/coprocessor/dag/expr/arithmetic.rs]: Yes, I didn't read full error message. Fixed.

**Comment by hicqu** [comment]
@BusyJay , PTAL, thanks.

File: src/coprocessor/dag/expr/arithmetic.rs [hunk 13]
```diff
@@ -0,0 +1,427 @@
...
+            (true, true) => (lhs as u64).checked_sub(rhs as u64).map(|t| t as i64),
+            (true, false) => if rhs >= 0 {
+                (lhs as u64).checked_sub(rhs as u64).map(|t| t as i64)
+            } else {
+                (lhs as u64).checked_add(opp_neg!(rhs)).map(|t| t as i64)
+            },
+            (false, true) => if lhs >= 0 {
+                (lhs as u64).checked_sub(rhs as u64).map(|t| t as i64)
+            } else {
+                None
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/arithmetic.rs]: Should return error here?

File: src/coprocessor/dag/expr/arithmetic.rs [hunk 14]
```diff
@@ -0,0 +1,427 @@
...
+        ctx: &StatementContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Decimal>>> {
+        let lhs = try_opt!(self.children[0].eval_decimal(ctx, row));
+        let rhs = try_opt!(self.children[1].eval_decimal(ctx, row));
+        let result: Result<Decimal> = lhs.mul(&rhs).into();
+        result.map(Cow::Owned).map(Some)
+    }
+
+    pub fn multiply_int(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/arithmetic.rs]: Should follow the same pattern as `plus_int` or `minus_int`.

File: src/coprocessor/dag/expr/arithmetic.rs [hunk 15]
```diff
@@ -0,0 +1,427 @@
...
+            }
+            Ok(Some(res))
+        } else {
+            Ok(Some(res as i64))
+        }
+    }
+}
+
+#[inline]
+fn check_integer_overflow(us: bool, res: i64, res_us: bool) -> Result<Option<i64>> {
```
Comment (BusyJay) [inline src/coprocessor/dag/expr/arithmetic.rs]: Can be removed now.

File: src/coprocessor/dag/expr/arithmetic.rs (hunk 14 above)
Comment (hicqu) [inline src/coprocessor/dag/expr/arithmetic.rs]: fixed.

**Pull request merged**